File: Data Structures & Algorithms/binary-search/submission-0.cs
public class Solution
{
	public int Search(int[] nums, int target)
	{
		
		int start = 0;
		int end = nums.Length - 1;
		
		while(start <= end)
		{
			int searchIndex = (start + end) / 2;
			int cur = nums[searchIndex];
			
			if(target == cur)
			{
				return searchIndex;
			}
			else if (target > cur)
			{
				// Focus on the right side of the array
				start = searchIndex + 1;
			}
			else
			{
				// Focus on the left side of the array
				end = searchIndex - 1;
			}
		}
		return -1;
    }
}

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-0.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        # Find the index of the smallest value by going right until the next value is larger than the current one.
      for i in range(len(nums) - 1, -1, -1):
        if nums[i - 1] > nums[i]:
          break
        
        # Perform a binary search on the left side first for target.
      left, right = 0, i - 1
      while left <= right:
        mid = left + (right - left) // 2
        if nums[mid] == target:
          return mid
        elif nums[mid] < target:
          left = mid + 1
        else:
          right = mid - 1
        
        # Perform a binary search on the right side for target.
      left, right = i, len(nums) - 1
      while left <= right:
        mid = left + (right - left) // 2
        if nums[mid] == target:
          return mid
        elif nums[mid] < target:
          left = mid + 1
        else:
          right = mid - 1

      return -1